### PR TITLE
websocket: use FixedQueue instead of Set

### PR DESCRIPTION
**Pull request by tsctx**
@KhafraDev

### CODE REVIEW

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by tsctx** [comment]
> Why not `fixed_queue` from node core?

There is no particular reason for this.

**Comment by ronag** [comment]
> In this use case, I prefer the simple linkedlist; Initialization of the array of 2048 entries takes time :)

I think that is amortized quite quickly...

**Comment by tsctx** [comment]
> do you have benchmarks?

What about?

**Comment by tsctx** [comment]
If it was about sending data, it would be 1.2 times faster since we changed to zero-copy.

**Comment by KhafraDev** [comment]
Yeah, sending messages. Do you have the data to backup the 1.2x faster claim?

**Comment by tsctx** [comment]
okey, look this

```
benchmark      time (avg)             (min … max)       p75       p99      p999
------------------------------------------------- -----------------------------
• send
------------------------------------------------- -----------------------------
undici      1'467 µs/iter   (1'187 µs … 6'165 µs)  1'442 µs  4'022 µs  6'165 µs <- before
undici      1'658 µs/iter   (1'305 µs … 4'145 µs)  1'673 µs  3'043 µs  4'145 µs <- after
```

**Comment by tsctx** [comment]
> > In this use case, I prefer the simple linkedlist; Initialization of the array of 2048 entries takes time :)
> 
> I think that is amortized quite quickly...

Rewrite with fixed_queue!

**Review by KhafraDev — APPROVED** [review]
why is queue being lazily created?

**Review by KhafraDev — CHANGES_REQUESTED** [review]
tests are failing

**Comment by tsctx** [comment]
I don't have time right now, so I will rewrite it in another PR.

File: lib/web/websocket/sender.js [hunk 1]
```diff
@@ -2,15 +2,30 @@
...
+ * @property {Buffer | null} frame
+ */
+
 class SendQueue {
-  #queued = new Set()
-  #size = 0
+  /**
+   * @type {FixedQueue | null}
+   */
+  #queue = null
```
Comment (KhafraDev) [inline lib/web/websocket/sender.js:21]:
```suggestion
  /**
   * @type {FixedQueue}
   */
  #queue = new FixedQueue()
```

File: lib/web/websocket/sender.js [hunk 2]
```diff
@@ -19,66 +34,78 @@ class SendQueue {
...
-        this.#run()
+        /** @type {SendQueueNode} */
+        const node = {
+          promise: null,
+          callback: cb,
+          frame
+        }
+        if (this.#queue === null) {
+          this.#queue = new FixedQueue()
+        }
```
Comment (KhafraDev) [inline lib/web/websocket/sender.js:50]:
```suggestion
```

File: lib/web/websocket/sender.js [hunk 3]
```diff
@@ -19,66 +34,78 @@ class SendQueue {
...
+        node.frame = createFrame(ab, hint)
+      }),
+      callback: cb,
+      frame: null
+    }
 
-      if (!done) return
+    if (this.#queue === null) {
+      this.#queue = new FixedQueue()
+    }
```
Comment (KhafraDev) [inline lib/web/websocket/sender.js:68]:
```suggestion
```

File: lib/web/websocket/sender.js [hunk 4]
```diff
@@ -19,66 +34,78 @@ class SendQueue {
...
-    const frame = new WebsocketFrameSend()
-    const opcode = hint === sendHints.string ? opcodes.TEXT : opcodes.BINARY
-
-    frame.frameData = data
-    const buffer = frame.createFrame(opcode)
-
-    this.#socket.write(buffer, cb)
+  async #run () {
+    this.#running = true
+    /** @type {FixedQueue} */
```
Comment (tsctx) [inline lib/web/websocket/sender.js:79]:
```suggestion
```

File: lib/web/websocket/sender.js [hunk 5]
```diff
@@ -19,66 +34,71 @@ class SendQueue {
...
+    const node = {
+      promise: item.arrayBuffer().then((ab) => {
+        node.promise = null
+        node.frame = createFrame(ab, hint)
+      }),
+      callback: cb,
+      frame: null
+    }
 
-      if (!done) return
```
Comment (tsctx) [inline lib/web/websocket/sender.js:63]:
```suggestion
```

**Comment by KhafraDev** [comment]
Thanks, WebSocket is bidirectional so it doesn't make sense to only be receiving messages (which is the only benefit I can see in lazily creating the queue?). We also have EventSource for that use case.

**Comment by tsctx** [comment]
When I received the message, I mistakenly thought I was generating a queue.

So there is no advantage.

**Review by KhafraDev — APPROVED** [review]

**Pull request merged**